client.go

```diff
@@ -87,39 +90,70 @@ func (c *Client) SetID(id string) error {
 	return nil
 }
 
-func (c *Client) Consumer(topics []string, cfg rdkafka.ConfigMap) (*Consumer, error) {
-	if !c.consReady {
-		return nil, errors.New("Connection not ready. Identify yourself using `CLIENT SETNAME` first")
-	}
+// registerConsumer registers a new Consumer denoted by cid.
+func (c *Client) registerConsumer(
+	cid ConsumerID, gid string, topics []string, cfg rdkafka.ConfigMap) (*Consumer, error) {
 
-	// we include the topics in the consumer's id so that GetOrCreate()
-	// creates a new consumer if the client issues a BLPOP for a different
-	// topic
-	cid := ConsumerID(fmt.Sprintf("%s|%s", c.id, strings.Join(topics, ",")))
+	// apparently, reusing the same config between consumers silently makes them non-operational
+	kafkaCfg := rdkafka.ConfigMap{}
+	for k, v := range c.cfg.Librdkafka.Consumer {
+		err := kafkaCfg.SetKey(k, v)
+		if err != nil {
+			return nil, err
+		}
+	}
+	for k, v := range cfg {
+		err := kafkaCfg.SetKey(k, v)
+		if err != nil {
+			return nil, err
+		}
+	}
+	err := kafkaCfg.SetKey("group.id", gid)
+	if err != nil {
+		return nil, err
+	}
 
-	if c.consID == "" {
-		c.consID = cid
+	// Extract the consumer name from the client id.
+	// We know by client.Consumer() that cid is in the form of `<group:name>|<topics>`
+	cidNoTopics := strings.Split(strings.Split(string(cid), "|")[0], ":")[1]
+	err = kafkaCfg.SetKey("client.id", cidNoTopics)
+	if err != nil {
+		return nil, err
 	}
 
-	if cid != c.consID {
-		return nil, fmt.Errorf("Client '%s' has already Consumer '%s' registered.", c.id, c.consID)
+	ctx, consumerCancel := context.WithCancel(c.ctx)
+	cons, err := NewConsumer(cid, topics, kafkaCfg)
+	if err != nil {
+		return nil, err
 	}
+	cons.cancel = consumerCancel
 
-	return c.consManager.GetOrCreate(cid, c.consGID, topics, cfg)
+	go cons.Run(ctx)
+
+	return cons, nil
 }
 
-// fetchConsumer returns the registered Consumer for the current Client.
-func (c *Client) fetchConsumer() (*Consumer, error) {
-	if c.consID == "" {
-		return nil, errors.New("No consumer is registered for Client " + c.id)
+func (c *Client) Consumer(topics []string, cfg rdkafka.ConfigMap) (*Consumer, error) {
```

Comment:
While we're at it, we should add some documentation to this method. I also believe it should be renamed to `GetOrCreateConsumer` or something.

Reply:
Actually, it should be `NewConsumer`.

Reply:
See, my comment [here](https://github.com/skroutz/rafka/pull/84#discussion_r328969782). Let's handle this scenario in another PR.